### ISSUE TITLE
Pinch distance seems to be calculated differently in Safari than in Chrome/Firefox

### ISSUE THREAD

**Issue by samselikoff** [pmndrs/use-gesture#260]
**Describe the bug**
The pinch gesture behaves differently on Safari than it does on Chrome/Firefox.

The initial zoom is the same, but after you zoom in, let go, then start zooming out, the distance calculated seems to be much smaller on Safari than on Chrome/Firefox.

**Sandbox or Video**
I'm using the demo David made for [my last issue](https://github.com/pmndrs/react-use-gesture/issues/255).

Fullscreen app (easiest to see here): https://oj203.csb.app
Sandbox: https://codesandbox.io/s/oj203

**Information:**
 - React Use Gesture version: Sandbox is 9.0.0-beta.11 (but tested on latest)
 - Device: Mac + Trackpad
 - Browser: Safari

**Comment by dbismut** [comment]
Hi Sam — good catch. It's a bit of a trick here, the lib uses `WebkitGestureEvent` on Safari (the only browser that supports it), also enabling rotation. This happens only when it can (using domTarget is a requirement since React doesn't have handlers for `WebkitGestureEvents`.

To normalise the values between events, we divide the zoom factor by an arbitrary constant that is `260` (there wasn't extensive testing regarding this value). However, I forgot to normalise the ctrl + wheel event, which I now did empirically by multiplying the wheel delta by `7`. This makes the pinch much faster on Chrome and Firefox, **but** zooming with wheel and ctrlKey will feel really really fast. Unfortunately there's no way that I know to determine when the user is using a trackpad or a mouse + ctrlKey. Both events are returned as `wheel` events with `{ ctrlKey: true }`.

Anyway, I've just released 9.0.2, let me know if it fixes it for you.

**Comment by samselikoff** [comment]
Thanks for looking into this!

Sadly looks like I'm seeing the same behavior on 9.0.2. Try the fullscreen demo on Chrome then Safari: https://137ns.sse.codesandbox.io

https://user-images.githubusercontent.com/2922250/107859054-2041b100-6e05-11eb-99b8-51dab1f6a9b1.mov

Here's the Codesandbox: https://codesandbox.io/s/floral-architecture-137ns?file=/pages/index.js

Let me know if there's anything else I'm missing :)

**Comment by dbismut** [comment]
Got it @samselikoff, this happened when dezooming while I focused on zooming! Hopefully this should be now fixed in `9.0.4`. It's hard to have a 1:1 behavior between all browsers since Safari uses different event, but that should be closer to the expected behavior!

**Comment by samselikoff** [comment]
Still seeing this:

![4BB24F09-EB59-4378-B3A0-E19D3C81BAB9_1_105_c](https://user-images.githubusercontent.com/2922250/108079728-21f5b980-703d-11eb-83cc-5100ecc876f4.jpeg)

It could be that my math is off. What I'd like to have happen, is for the same distance to zoom out the image, whether I've lifted up my fingers or not. If that makes sense.

**Comment by samselikoff** [comment]
And that does seem to be how Chrome works too. So I'm not sure if it's a math thing or a browser thing.

**Comment by dbismut** [comment]
That's not your math there's indeed something wrong here. Let me investigate, it's a bit harder than it looks, I'm not quite sure how to fix this yet :S

**Comment by dbismut** [comment]
Hi @samselikoff this took a while to solve, and the API is ugly at this point, I'll release something better in the next stable version, but you can install `9.1.0-alpha.1`.

The correct way to calculate distance is this one:

```js
onPinch: ({ event, initial: [id], da: [d], offset: [, a], memo = scale.get() }) => {
  const scale = (memo * d) / id
  set({ scale, rotateZ: a })
  return memo
}
```

This actually makes more sense compared to how I was using distance previously. The scale of an object should be its previous scale multiplied by the current scale, whereas before I was doing some sums which worked ok when the scale started with 1 but got crazy otherwise. I'm surprised nobody picked this before you 😅

The good part is that it stops using an arbitrary value, which I didn't like anyway. The bad part is that it forces you to use memo, and make movement and offset useless for scaling (this is what I want to fix).

Obviously this has some impact on your project, as it has on mine. I've just updated it. The code is a bit cleaner IMO, and I'll also update the Exaclidraw with formulas that make sense.

Have a look here:
https://codesandbox.io/s/elated-lichterman-oj203

Let me know if it works out. If yes, I'll release a stable version next week. I'll probably remove the idea of distance and only keep scale which is I guess what 99% people are using anyway!

**Comment by dbismut** [comment]
@samselikoff Hey curious to know if this fixes it for you.

**Comment by samselikoff** [comment]
Ok trying this out! Struggling a bit translating it to my code. What is `da`?

**Comment by dbismut** [comment]
da plays the same role as xy for drag, it's the raw value of the distance between fingers. Essentially the scale factor is the current distance between your fingers divided by the initial distance between your fingers.

But again I'll be working on a full refactor of this to make it cleaner. I just want to know if this works out before proceeding.

**Comment by samselikoff** [comment]
Gotcha. Ok I'm logging it out and it always seems to start at 60. Does that seem correct?

**Comment by dbismut** [comment]
Yes this is a hack on trackpads to make it work properly. If I redo the implementation in v10 I'd probably try not to expose this.

**Comment by samselikoff** [comment]
Gotcha. Should I account for it in my calculations then? Subtract 60?

**Comment by dbismut** [comment]
Nope, since scale is a fraction you don't need to subtract anything.

```js
useDrag( ({ da: [d], initial:[id] }) => {
  const scale = d / id
})
```

**Comment by samselikoff** [comment]
Getting closer. Having trouble with the pinch origin correction logic, and I actually just noticed it in your sandbox as well. Try zooming in on the corner of your image, stopping, and starting a new pinch-out gesture exactly the same place. The image will zoom out about the wrong origin.

I think it has to do with the fact that we're using percentages.

```js
let scale = d / id
```

On the first (pinch-in) gesture, say this goes from 1.0 to 1.5.

On the second (pinch-out) gesture, this will go from 1.0 down to (say) .6. And if you multiply it by the starting scale of 1.5, the math doesn't bring you back.

1.0 * 1.5 = 150% (ending percent)
.6 * 1.5 = 90% (ending percent)

Let me know if it makes sense! I think we might need to back out the pixels moved to properly calculate the change? Or, can `d` and `id` be "remembered" the way offset is?

**Comment by dbismut** [comment]
Hi @samselikoff I think this is just a silly omission from my end, where I forgot to fix on the sandbox what I was testing locally. I think it's now working properly.

But essentially, I had:

```js
// before
const tx = ox - (x + width / 2) / initialScale
const ty = oy - (y + height / 2) / initialScale

// after
const tx = ox - (x + width / 2)
const ty = oy - (y + height / 2)
```

Dividing by `initialScale` doesn't make sense in that case, since we're already accounting for the `initialScale` in the end scale. I've tried to fix the excalidraw:

https://excalidraw.com/#json=5404749805387776,4ighJ-_I-Os_lvDbyWPMPg

Sorry for the back and forth here.

**Comment by samselikoff** [comment]
Wew, that did it!!

Last wrinkle. `d` and `id` don't respect `distanceBounds: { min: 0 }`. So if I add logic like this to my drag handler

```js
if (newScale >= 1) {
  scale.set(newScale);
  x.set(newX);
  y.set(newY);
} else {
  scale.set(1);
}
```

and the user keeps zooming out, `d` and `id` will keep going down. Then if the user zooms in again it's more sensitive, due to the values. Is there a way to "reset" or floor these values?

**Comment by dbismut** [comment]
At this point no, I'd have to get deep in the internals to make it work properly and make the API consistent across hooks... will try to have a shot at it next week.

**Comment by samselikoff** [comment]
Gotcha. Can you think of any workarounds for now?

**Comment by dbismut** [comment]
Nothing super simple I'm afraid, you can clamp the scale, but the problem is that `da[0]` continues to change while your scale is set to its minimum.

You'd be tempted to do the below:

```js
const scale  = Math.max(0.25, : memo[4] * d / id)
// This will work for the first gesture but then will break.
// you need to ensure continuity between d and take over from
// where you left it.
const ms = scale / memo[4]
const x = memo[0] - (ms - 1) * memo[2]
const y = memo[1] - (ms - 1) * memo[3]
set({ scale, rotateZ: a, x, y })
```

Look at the above, let's say that your minimum scale is `0.25`.
- Let's say that `memo[4]` (`scale.get()`) starts at `1`.
- During the first gesture, we zoom out.
- `id` starts at 60 (initial distance), and `d` will decrease `60` to `6`.
- The scale without being clamped should be `1 * 6 / 60 = 0.1`
- But here it's clamped to `0.25`, which is like if would have stopped at `d = 15`. All good for now.
- Now start the gesture again, this time zoom in.
- `d` will increase from `6` to `60`. `id` this time is equal to `6` (last value of `d`). `memo[4]` is equal to `0.25`. 
- Based on the math above, the new scale should be: `0.25 * 60 / 6 = 2.5`.

But that's way too much, in reality you would have want to store the value from `d` matching the `0.25` scale, so `d=15`. Starting with `15` would  have resulted in the following scale: `0.25 * 69 / 15 = 1,15`.

So what you need to do is store the value of `d` and swap `id` with your own values. This will look ugly in the code, but that's achievable.

**Comment by samselikoff** [comment]
I see. Thanks for digging into this (again)!

I may try to do this but might wait in case you choose to expose something new. I'll keep my eye out but feel free to close this issue! And happy to try out anything you find to verify it works in my use case.

**Comment by samselikoff** [comment]
Looks like the 60 value was removed in my upgrade from 9.1.0-alpha.1 to 9.1.3! Cool except now I'm a little confused how to calculate distance.

Before I was doing

```js
let movementScaled = distance / initialDistance
```

but now `initialDistance` starts at 0 so I get a divide by 0 error.

Should it be something like `(distance - initialDistance)` now? Getting a little confused because I've stepped away from this for a bit.

Or should I go back to using `offset`?

**Comment by dbismut** [comment]
Hi @samselikoff this is just terrible version numbering. The 9.1.0-alpha is actually the only one properly handling pinch. I though I would be able to integrate it on 9.1 but this is way more effort than I thought, so I'll keep this for v10. In other words you should keep using 9.1.0-alpha.1. 

Sorry about this.

**Comment by samselikoff** [comment]
No problem at all! Let me see if I can revert + get it working 👍

**Comment by samselikoff** [comment]
I reverted and got the original behavior back, thanks so much for chiming in!

The last quirk I'm seeing on mobile is, if I start pinching, and then I also move my fingers in a way that's interpreted as a drag while pinching, the dragging is inverted. That is if I'm pinching out (so the picture is zooming in) and also moving my fingers down and to the right, the image zooms in (correctly) but starts moving up and to the left (opposite).

Have you seen this behavior before?

**Comment by dbismut** [comment]
Closing as fixed in v10 #300. Please let me know if the issue persists somehow.

**Comment by samselikoff** [comment]
ok nice work man!! Looking forward to trying this

**Comment by ullatap** [comment]
I am using v10.0.0 and I am still getting the same behavior @samselikoff described above. Is there any update or a workaround for this issue?

**Comment by dbismut** [comment]
Reopening @ullatap, will release a fix. Sorry about this, now Safari is showing the right behavior, and Chrome / Firefox aren't (you're actually experiencing the **opposite** problem @samselikoff had^^).

**Comment by dbismut** [comment]
@ullatap I've given a stab at it on PR #354.

Just a bit of context: Safari with magic trackpads uses WebkitGestureEvents which provide `scale` and `rotation` as attributes (most probably Safari actually calculates these properties as it knows the actual positions of the fingers on the trackpad). However, Chrome and Firefox interpret trackpad pinch as a wheel event with the ctrl key pressed (this was quite of a surprise to me when I heard about it). Therefore the only property I get from the `WheelEvent` is `deltaY` and I'm trying to sort of make it work like a scale property.

So long story short, it's impossible to get the very same behavior. With this PR, I've tried to multiply `deltaY` by the initial scale of the object, and it gave satisfactory results on my MBA M1 on all tested browsers (Firefox, Chrome, Safari TP).

Here's a sandbox you may want to try, let me know how it works out for you. If you find it ok, then I'll make a patch release.
https://codesandbox.io/s/gesture-pinch-d9seu

**Comment by ullatap** [comment]
@dbismut thank you for the fast reply and the explanation. I tested the sandbox version in Chrome, Firefox and Safari and it works well in all 3 (no sensitivity issues detected). :)

To clarify, the behavior I detected in my app was that in Chrome (and Samsung Internet browser) pinching was over-sensitive at first but after zooming in unzooming became very slow. Hope this can be fixed with the patch, I find the app otherwise very good. :)

**Comment by dbismut** [comment]
Oh ok. So just so that I'm sure I understand, the bug you're facing is experienced on an Android device right?

**Comment by dbismut** [comment]
Oh ok. So just so that I'm sure I understand, the bug you're facing is experienced on an Android device right?

**Comment by ullatap** [comment]
Yes, on Samsung Galaxy.

**Comment by dbismut** [comment]
Does that sandbox behave correctly on your device? https://codesandbox.io/s/gesture-pinch-d9seu
I've tried Chrome on an Android phone and it works like it should I think.

**Comment by ullatap** [comment]
Yes, tested it on Chrome on my Android and it works great.

**Comment by rnarcos** [comment]
Hi, 👋

I'm not entirely sure that this has been fixed.

I've followed the tutorial from @samselikoff's video on how to correct pinch to zoom origin (Thanks for that, by the way), but I am having different experiences on the following devices / browsers:

https://user-images.githubusercontent.com/35005829/209559215-2b8102b4-83d5-40d9-9a6b-7475a4bcd03e.mov

On macOS Monterey Version 12.6's chrome (Version 108.0.5359.124 (Official Build) (arm64)) - on the browser itself and on the simulator:
✅: Pinch gesture zooms in without effort;
✅: After-pinch gesture resets the animations to their initial state;


https://user-images.githubusercontent.com/35005829/209559198-273a081a-e218-4013-9275-9e2b57ec1eb3.mov

On macOS Monterey Version 12.6's safari (Version 16.0 (17614.1.25.9.10, 17614)):
❌: Pinch gesture zooms in without effort;
❌: After-pinch gesture resets the animations to their initial state;


https://user-images.githubusercontent.com/35005829/209559860-183b1227-d64a-4a32-ac0d-a95e1f0ca81e.mov

On iOs 16.1's chrome (Version 108.0.5359.112):
❌: Pinch gesture zooms in without effort;
✅: After-pinch gesture resets the animations to their initial state;

https://user-images.githubusercontent.com/35005829/209559862-64cfa371-f105-47b6-bcf3-cd4d58a2079f.mov

On iOS 16.1's safari:
❌: Pinch gesture zooms in without effort;
✅: After-pinch gesture resets the animations to their initial state;

I'm using a slightly different approach than Sam's video suggests, because my requirement specified that zoom should be reseted after user has stopped pinching, so this is what my `useGesture` looks like:

```typescript
useGesture(
  {
    onPinch: ({
      memo,
      origin: [pinchOriginX, pinchOriginY],
      offset: [d],
      pinching,
    }) => {
      if (!pinching) return;

      memo ??= {
        bounds: memo?.bounds ?? ref.current!.getBoundingClientRect(),
        transform: transform,
      } as Memo;

      const transformOriginX = memo.bounds.x + memo.bounds.width / 2;
      const transformOriginY = memo.bounds.y + memo.bounds.height / 2;

      const displacementX =
        (transformOriginX - pinchOriginX) / memo.transform.scale;
      const displacementY =
        (transformOriginY - pinchOriginY) / memo.transform.scale;

      const initialOffsetDistance = (memo.transform.scale - 1) * 50;
      const movementDistance = d - initialOffsetDistance;

      setTransform({
        x: memo.transform.x + (displacementX * movementDistance) / 50,
        y: memo.transform.y + (displacementY * movementDistance) / 50,
        scale: 1 + d / 50,
      });

      return memo;
    },
    onWheelEnd: resetTransform,
    onTouchEnd: resetTransform,
  },
  {
    target: ref,
    eventOptions: { passive: false },
    pinch: {
      pinchOnWheel: true,
      from: () => [1, 0],
    },
  },
);
```

I don't exactly understand why macOS safari's do not trigger the `resetTransform` at the `onWheelEnd` and `onTouchEnd` handlers, I also don't fully understand if I'm using the `from` property correctly - would love your input on that one @dbismut!

Awesome work, guys!

**Comment by dbismut** [comment]
@rnarcos the library has evolved a bit since, and you might want to try the algorithm in this sandbox:

https://codesandbox.io/s/github/pmndrs/use-gesture/tree/main/demo/src/sandboxes/card-zoom

If this doesn't work, feel free to ping me on Discord and we'll try to solve this together.
